streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -340,32 +355,67 @@ public SegmentSearchResult find(final long timestamp, final boolean includeValue
         }
 
         @Override
-        public List<SegmentSearchResult> findAll(final long fromTime, final long toTime) {
-            long currNextTimestamp = nextTimestamp;
-            final List<SegmentSearchResult> segmentSearchResults = new ArrayList<>();
-            long currTimestamp = -1L; // choose an invalid timestamp. if this is valid, this needs to be re-worked
-            int currValueSize;
-            int currIndex = 0;
-            int cumValueSize = 0;
-            while (currTimestamp != minTimestamp) {
-                final int timestampSegmentIndex = 2 * TIMESTAMP_SIZE + currIndex * (TIMESTAMP_SIZE + VALUE_SIZE);
+        public SegmentSearchResult find(final long fromTime, final long toTime, final ResultOrder order) {
+            // this segment does not have any record in query specified time range
+            if (toTime < minTimestamp || fromTime > nextTimestamp) {
+                return null;
+            }
+            long currTimestamp = -1;
+            long currNextTimestamp = -1;
+
+
+            if (order.equals(ResultOrder.ASCENDING) && valuesStartingIndex == -1) {
+                findValuesStartingIndex();
+            }
+
+            while (hasStillRecord(currTimestamp, currNextTimestamp, order)) {
+                final int timestampSegmentIndex = getTimestampIndex(order, currentDeserIndex);
                 currTimestamp = ByteBuffer.wrap(segmentValue).getLong(timestampSegmentIndex);
-                currValueSize = ByteBuffer.wrap(segmentValue).getInt(timestampSegmentIndex + TIMESTAMP_SIZE);
-                cumValueSize += Math.max(currValueSize, 0);
+                currNextTimestamp = timestampSegmentIndex == 2 * TIMESTAMP_SIZE ? nextTimestamp // if this is the first record metadata (timestamp + value size)
+                                                                                : ByteBuffer.wrap(segmentValue).getLong(timestampSegmentIndex - (TIMESTAMP_SIZE + VALUE_SIZE));
```

Comment:
Cf other comment blow: can we avoid parsing to get back the long, but would we get it fom a cache?